src/twisted/mail/test/test_smtp.py

```diff
@@ -674,6 +676,50 @@ def check(ignored):
 
         return self.loopback(server, client).addCallback(check)
 
+    def test_ESMTPSenderFactory_TLSError(self):
+        """
+        Attempting to connect to an ESMTP server which presents an invalid certificate
+        should produce a meaningful error.
+        """
+        # set up a dummy ESMTP server which will present a self-signed cert after
+        # STARTTLS
+        buildServerProtocol = lambda: DummyESMTP(contextFactory=ServerTLSContext())
+        serverFactory = Factory.forProtocol(buildServerProtocol)
+        serverPort = reactor.listenTCP(0, serverFactory, interface="127.0.0.1")
+        self.addCleanup(serverPort.stopListening)
+
+        # build a client, which won't accept the certificate presented by the dummy
+        # server.
+        sentDeferred = defer.Deferred()
+        clientFactory = smtp.ESMTPSenderFactory(
+            "username",
+            "password",
+            "source@address",
+            "recipient@address",
+            BytesIO(b"message"),
+            sentDeferred,
+            retries=0,
+            requireAuthentication=False,
+            contextFactory=optionsForClientTLS("testdomain"),
+        )
+
+        # connect the two together
+        connector = reactor.connectTCP(
+            serverPort.getHost().host, serverPort.getHost().port, clientFactory
+        )
+        self.addCleanup(connector.disconnect)
+
+        def errback(f):
+            # check that we got an exception with a plausible error code.
+            f.trap(smtp.SMTPConnectError)
+            self.assertIn("certificate verify failed", str(f.value))
+
+        def callback(_res):
+            raise Exception("ESMTPSenderFactory unexpectedly completed")
```

Comment:
Codecov is unhappy about this line never being exercised, which is giving you a ❌ .  To pacify it, you might consider using `self.assertFailure` (whose documentation sadly omits that it returns the `Exception` object), like this:

```python3
def checkVerifyFailedMessage(e: smtp.SMTPConnectError):
    self.assertIn("certificate verify failed", str(e))
return self.assertFailure(sentDeferred, smtp.SMTPConnectError).addCallback(checkVerifyFailedMessage)
```

This leaves the error reporting to the testing framework rather than inlining it within the test where it's going to not be exercised and leave us coverage gaps.

Reply:
done, thanks for the tip.